packages/horizon/contracts/staking/HorizonStaking.sol

```diff
@@ -923,20 +985,30 @@ contract HorizonStaking is HorizonStakingBase, IHorizonStakingMain {
         }
 
         // decode
-        (uint256 tokensThawed, uint256 tokensThawing, uint256 sharesThawing) = abi.decode(
+        (uint256 tokensThawed, uint256 tokensThawing, uint256 sharesThawing, uint256 thawingNonce) = abi.decode(
             _acc,
-            (uint256, uint256, uint256)
+            (uint256, uint256, uint256, uint256)
         );
 
-        // process
-        uint256 tokens = (thawRequest.shares * tokensThawing) / sharesThawing;
-        tokensThawing = tokensThawing - tokens;
-        sharesThawing = sharesThawing - thawRequest.shares;
-        tokensThawed = tokensThawed + tokens;
-        emit ThawRequestFulfilled(_thawRequestId, tokens, thawRequest.shares, thawRequest.thawingUntil);
+        // process - only fulfill thaw requests for the current valid nonce
+        uint256 tokens = 0;
+        bool invalidThawRequest = thawRequest.thawingNonce != thawingNonce;
+        if (!invalidThawRequest) {
```

Comment:
micronit: we're always using`!invalidThawRequest`, so might be slightly cleaner to use a `validThawRequest` instead?